src/station_cmd.cpp

```diff
@@ -4014,68 +4014,68 @@ const StationList *StationFinder::GetStations()
+		assert(st->owner <= OWNER_NONE);
 
-		if (IsCargoInClass(type, CC_PASSENGERS)) {
-			if (st->facilities == FACIL_TRUCK_STOP) continue; // passengers are never served by just a truck stop
-		} else {
-			if (st->facilities == FACIL_BUS_STOP) continue; // non-passengers are never served by just a bus stop
-		}
-
-		/* This station can be used, add it to st1/st2 */
-		if (st1 == nullptr || st->goods[type].rating >= best_rating1) {
-			st2 = st1; best_rating2 = best_rating1; st1 = st; best_rating1 = st->goods[type].rating;
-		} else if (st2 == nullptr || st->goods[type].rating >= best_rating2) {
-			st2 = st; best_rating2 = st->goods[type].rating;
+		byte r = st->goods[type].rating;
+		if (r > company_best[st->owner]) {
+			best_sum += r - company_best[st->owner];  // it's usually faster than iterating companies later
+			company_best[st->owner] = r;
+			any_station = st;
+			if (r > best_rating) best_rating = r;
 		}
+		company_sum[st->owner] += r;
 	}
 
 	/* no stations around at all? */
-	if (st1 == nullptr) return 0;
+	if (best_rating == 0) return 0;
 
 	/* From now we'll calculate with fractal cargo amounts.
 	 * First determine how much cargo we really have. */
-	amount *= best_rating1 + 1;
+	amount *= best_rating + 1;
 
-	if (st2 == nullptr) {
+	if (best_sum == best_rating && company_sum[any_station->owner] == best_rating) {
 		/* only one station around */
-		return UpdateStationWaiting(st1, type, amount, source_type, source_id);
-	}
-
-	/* several stations around, the best two (highest rating) are in st1 and st2 */
-	assert(st1 != nullptr);
-	assert(st2 != nullptr);
-	assert(best_rating1 != 0 || best_rating2 != 0);
-
-	/* Then determine the amount the worst station gets. We do it this way as the
-	 * best should get a bonus, which in this case is the rounding difference from
-	 * this calculation. In reality that will mean the bonus will be pretty low.
-	 * Nevertheless, the best station should always get the most cargo regardless
-	 * of rounding issues. */
-	uint worst_cargo = amount * best_rating2 / (best_rating1 + best_rating2);
-	assert(worst_cargo <= (amount - worst_cargo));
-
-	/* And then send the cargo to the stations! */
-	uint moved = UpdateStationWaiting(st1, type, amount - worst_cargo, source_type, source_id);
-	/* These two UpdateStationWaiting's can't be in the statement as then the order
-	 * of execution would be undefined and that could cause desyncs with callbacks. */
-	return moved + UpdateStationWaiting(st2, type, worst_cargo, source_type, source_id);
+		return UpdateStationWaiting(any_station, type, amount, source_type, source_id);
+	}
+
+	uint moving = 0;
+	for (Station *st : *all_stations) {
+		if (!CanMoveGoodsToStation(st, type)) continue;
+		/* Ignore rounding loss as it's only 1/256 of unit per station in the worst case */
+		uint station_amount = amount * company_best[st->owner] * st->goods[type].rating / best_sum / company_sum[st->owner];
```

Comment:
It's not immediately obvious this expression won't allow more than the amount of cargo to distribute to be distributed. Can you write something explaining how the sum of all these fractions never exceeds 1?